src/Oscoin/Deployment/Internal/CloudInit.hs

```diff
@@ -0,0 +1,241 @@
+{-# LANGUAGE DataKinds             #-}
+{-# LANGUAGE DuplicateRecordFields #-}
+
+module Oscoin.Deployment.Internal.CloudInit
+    ( CloudConfig
+    , mkCloudConfig
+    , renderCloudConfig
+
+    , User
+    , mkUser
+    , rootUser
+    , renderUidOrName
+
+    , Group
+    , mkGroup
+    , rootGroup
+    , renderGidOrName
+
+    , Owner
+    , mkOwner
+    , rootOwner
+
+    , WriteFiles
+    , mkWriteFiles
+
+    , FileContents
+    , plainTextFileContents
+    , base64FileContents
+
+    , FileMode
+    , fromPosixFileMode
+    )
+where
+
+import           Oscoin.Prelude hiding (group)
+
+import           Data.Aeson (ToJSON(..), (.=))
+import qualified Data.Aeson.Types as Aeson
+import           Data.ByteString.BaseN (AtBase, encodedText)
+import           Data.Generics.Product
+import qualified Data.Set as Set
+import           Data.Vector (Vector)
+import qualified Data.Yaml as Yaml
+import           Lens.Micro (set)
+import           Lens.Micro.Extras (view)
+import           Numeric (showIntAtBase)
+import qualified System.Posix as Posix
+
+data CloudConfig = CloudConfig
+    { users       :: Vector User
+    , groups      :: Vector Group
+    , write_files :: Vector WriteFiles
+    , runcmd      :: Vector Text
+    , bootcmd     :: Vector Text
+   } deriving Generic
+
+instance ToJSON CloudConfig where
+    toJSON cc = Aeson.object
+        [ "users"       .= users cc
+        , "groups"      .= map (view (the @"name")) (view (the @"groups") cc)
+        , "write_files" .= write_files cc
+        , "runcmd"      .= runcmd cc
+        , "bootcmd"     .= bootcmd cc
+        ]
+
+    toEncoding cc = Aeson.pairs $
+           "users"       .= users cc
+        <> "groups"      .= map (view (the @"name")) (view (the @"groups") cc)
+        <> "write_files" .= write_files cc
+        <> "runcmd"      .= runcmd cc
+        <> "bootcmd"     .= bootcmd cc
+
+instance Semigroup CloudConfig where
+    a <> b = CloudConfig
+        { users       = on (<>) users a b
+        , groups      = on (<>) (groups :: CloudConfig -> Vector Group) a b
+        , write_files = on (<>) write_files a b
+        , runcmd      = on (<>) runcmd a b
+        , bootcmd     = on (<>) bootcmd a b
+        }
+
+instance Monoid CloudConfig where
+    mempty  = CloudConfig mempty mempty mempty mempty mempty
+    mappend = (<>)
+
+mkCloudConfig :: CloudConfig
+mkCloudConfig = mempty
+
+renderCloudConfig :: CloudConfig -> Text
+renderCloudConfig cc = "#cloud-config\n\n" <> toS (Yaml.encode cc)
+
+data User = User
+    { name   :: Text
+    , uid    :: Maybe Word16
+    , groups :: Set Group
+    } deriving (Eq, Ord, Generic)
+
+instance ToJSON User where
+    toJSON u = Aeson.object $ catMaybes
+        [ Just $ "name" .= view (the @"name") u
+        , ("uid" .=) . toJSON <$> uid u
+        , if Set.null (view (the @"groups") u) then
```

Comment:
I know that this is an internal tool and it's pointless to antagonise on it (plus, I do understand that you were having some good fun with `generic-lens`) but I find interesting how you went for `view (the @"groups") u` rather than the more succinct `groups u` here 😉

Reply:
I am using `DuplicateRecordFields` everywhere due to lack of imagination. That doesn't always work well, but inference didn't fail me so far when using lenses.